url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/utils/Report.java

```diff
@@ -0,0 +1,386 @@
+		List<DBProtocol> protocols = new LinkedList<>();
+		BufferedReader reader = new BufferedReader(new FileReader(f));
+		String regex = "MP-\\{(\\d+), (\\d\\.\\d+)\\}";
+		Pattern pattern = Pattern.compile(regex);
+		String retrievedText = null;
+		for (String line = reader.readLine(); line  != null; line = reader.readLine()){
+			Matcher m = pattern.matcher(line);
+			while (m.find()) {
+				MunillaAndPeinadoProtocol p = new MunillaAndPeinadoProtocol(Double.parseDouble(m.group(2)));
+				p.setNumberOfRounds(Integer.parseInt(m.group(1)));
+				
+				protocols.add(p);
+			}
+		}
+		reader.close();
+		return protocols;
+	}
+
+	
+	private static List<DBProtocol> getMADProtocol(File f) throws NumberFormatException, IOException {
+		List<DBProtocol> protocols = new LinkedList<>();
+		BufferedReader reader = new BufferedReader(new FileReader(f));
+		String regex = "MAD-\\{(\\d+)\\}";
+		Pattern pattern = Pattern.compile(regex);
+		String retrievedText = null;
+		for (String line = reader.readLine(); line  != null; line = reader.readLine()){
+			Matcher m = pattern.matcher(line);
+			while (m.find()) {
+				retrievedText = m.group(1);
+				MADProtocol p = new MADProtocol();
+				p.setNumberOfRounds(Integer.parseInt(retrievedText));
+				protocols.add(p);
+			}
+		}
+		reader.close();
+		return protocols;
+	}
+
+	private static List<DBProtocol> getKimAndAvoineProtocol(File f) throws NumberFormatException, IOException {
+		List<DBProtocol> protocols = new LinkedList<>();
+		BufferedReader reader = new BufferedReader(new FileReader(f));
+		String regex = "KA-\\{(\\d+), (\\d\\.\\d+)\\}";
+		Pattern pattern = Pattern.compile(regex);
+		String retrievedText = null;
+		for (String line = reader.readLine(); line  != null; line = reader.readLine()){
+			Matcher m = pattern.matcher(line);
+			while (m.find()) {
+				KimAndAvoineProtocol p = new KimAndAvoineProtocol(Double.parseDouble(m.group(2)));
+				p.setNumberOfRounds(Integer.parseInt(m.group(1)));
+				
+				protocols.add(p);
+			}
+		}
+		reader.close();
+		return protocols;
+	}
+
+	private static List<DBProtocol> getHanckeAndKuhnProtocol(File f) throws NumberFormatException, IOException {
+		List<DBProtocol> protocols = new LinkedList<>();
+		BufferedReader reader = new BufferedReader(new FileReader(f));
+		String regex = "HK-\\{(\\d+)\\}";
+		Pattern pattern = Pattern.compile(regex);
+		String retrievedText = null;
+		for (String line = reader.readLine(); line  != null; line = reader.readLine()){
+			Matcher m = pattern.matcher(line);
+			while (m.find()) {
+				retrievedText = m.group(1);
+				HanckeAndKuhnProtocol p = new HanckeAndKuhnProtocol();
+				p.setNumberOfRounds(Integer.parseInt(retrievedText));
+				protocols.add(p);
+			}
+		}
+		reader.close();
+		return protocols;
+	}
+
+	private static List<DBProtocol> getBussardAndBaggat(File f) throws NumberFormatException, IOException {
+		List<DBProtocol> protocols = new LinkedList<>();
+		BufferedReader reader = new BufferedReader(new FileReader(f));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 335, 339. The resource is closed at line 348. There are other execution paths that do not contain closure statements, for example, when Integer.parseInt throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).